src/blant.c

```diff
@@ -575,34 +587,74 @@ int RunBlantInThreads(int k, int numSamples, GRAPH *G)
         for(i=0; i<_numCanon; i++) for(j=0; j<_numCanon; j++) _graphletDistributionTable[i][j] = 0;
     }
 
-    if(_THREADS == 1)
-	return RunBlantFromGraph(k, numSamples, G);
+    int indexSamplingRange[2];
+    if(_THREADS == 1) {
+        if (_sampleMethod == SAMPLE_INDEX) {
+            indexSamplingRange[0] = 0;
+            indexSamplingRange[1] = G->n - 1;
+        }
+        return RunBlantFromGraph(k, numSamples, G, _sampleMethod == SAMPLE_INDEX ? indexSamplingRange : NULL);
+    }
 
-    if (_sampleMethod == SAMPLE_INDEX)
-        Fatal("The sampling method '-s INDEX' does not yet support multithreading (feel free to add it!)");
+    // At this point, _THREADS must be greater than 1.
 
+    //
+    char childrenRunInfinitely = (_outputMode == indexMotifs
+            ||  _outputMode == indexMotifOrbits
+            ||  _outputMode == indexOrbits
+            ||  _outputMode == indexGraphlets
+    )
+    && _sampleMethod != SAMPLE_INDEX
+    && !_window;
 
-    // At this point, _THREADS must be greater than 1.
-    int samplesPerThread = numSamples/_THREADS;  // will handle leftovers later if numSamples is not divisible by _THREADS
 
+    // determine the numSamples per thread. -1 indicates that the child processes will run infinitely until a stop signal from parent is received
+    // Leftovers will be handled later if numSamples is not divisible by _THREAD
+    int samplesPerThread = childrenRunInfinitely ? -1 : numSamples/_THREADS;
+
+    int startNodesPerThread = G->n / _THREADS;
     FILE *fpThreads[_THREADS]; // these will be the pipes reading output of the parallel blants
+
     for(i=0;i<_THREADS;i++) {
         _THREAD_NUM = i;
-        fpThreads[i] = ForkBlant(_k, samplesPerThread, G);
+        if (_sampleMethod == SAMPLE_INDEX) {
```

Comment:
Aha. Now I finally see what this indexSamplingRange is for. However, this will not produce very effective parallelism because different nodes will require vastly different amounts of CPU time. We'll probably need to figure out a better way to do parallelism in this case. We'll need to be able to *predict* how much effort it'll take to create the index for each node, and then figure out a way to load-balance across parallel threads. In fact... it probably makes more sense to fork off 1 child blant for EACH node, but ensure no more than THREADS of them run at the same time. This'll be easy to do using my "parallel" script, if you generate the list of commands to run (G->n commands), and fire them off to my "parallel -THREADS" program.

Reply:
Can I get the script? And to use the script this way I need to add a BLANT option to run -sINDEX mode on only one start node. Is my understanding correct?